intel-sgx/async-usercalls/src/io_bufs.rs

```diff
@@ -0,0 +1,324 @@
+        let (_, write_offset) = self.offsets();
+        let rem = self.remaining_capacity();
+        let can_write = cmp::min(buf.len(), rem);
+        let end = cmp::min(self.buf_len, write_offset + can_write);
+        let n = end - write_offset;
+        unsafe {
+            let userbuf = &mut *self.userbuf.get();
+            userbuf[write_offset..write_offset + n].copy_from_enclave(&buf[..n]);
+        }
+        self.advance_write(n);
+        n + if n < can_write { self.write(&buf[n..]) } else { 0 }
+    }
+
+    /// This function returns a slice of bytes appropriate for writing to a socket.
+    /// Once some or all of these bytes are successfully written to the socket,
+    /// `self.consume()` must be called to actually consume those bytes.
+    ///
+    /// Returns None if the buffer is empty.
+    ///
+    /// Panics if called more than once in a row without either calling `consume()`
+    /// or dropping the previously returned buffer.
+    pub fn consumable_chunk(&mut self) -> Option<UserBuf> {
+        assert!(
+            Arc::strong_count(&self.userbuf) == 1,
+            "called consumable_chunk() more than once in a row"
+        );
+        let range = match self.offsets() {
+            (_, _) if self.read == self.write => return None, // empty
+            (r, w) if r < w => r..w,
+            (r, _) => r..self.buf_len,
+        };
+        Some(UserBuf(UserBufKind::Shared {
+            user: self.userbuf.clone(),
+            range,
+        }))
+    }
+
+    /// Mark `n` bytes as consumed. `buf` must have been produced by a call
+    /// to `self.consumable_chunk()`.
+    /// Panics if:
+    /// - `n > buf.len()`
+    /// - `buf` was not produced by `self.consumable_chunk()`
+    ///
+    /// This function is supposed to be used in conjunction with `consumable_chunk()`.
+    pub fn consume(&mut self, buf: UserBuf, n: usize) {
+        assert!(n <= buf.len());
+        const PANIC_MESSAGE: &'static str = "`buf` not produced by self.consumable_chunk()";
+        let buf = buf.into_shared().expect(PANIC_MESSAGE);
+        assert!(Arc::ptr_eq(&self.userbuf, &buf), "{}", PANIC_MESSAGE);
+        drop(buf);
+        assert!(Arc::strong_count(&self.userbuf) == 1, "{}", PANIC_MESSAGE);
+        self.advance_read(n);
+    }
+
+    fn len(&self) -> usize {
+        match self.offsets() {
+            (_, _) if self.read == self.write => 0,                      // empty
+            (r, w) if r == w && self.read != self.write => self.buf_len, // full
+            (r, w) if r < w => w - r,
+            (r, w) => w + self.buf_len - r,
+        }
+    }
+
+    fn remaining_capacity(&self) -> usize {
+        let len = self.len();
+        debug_assert!(len <= self.buf_len);
+        self.buf_len - len
+    }
+
+    fn offsets(&self) -> (usize, usize) {
+        (self.read as usize % self.buf_len, self.write as usize % self.buf_len)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.read == self.write
```

Comment:
Shouldn't the modulo of these offsets be taken here as well?

Reply:
Using `len()` would keep most of these details in the `len()` and `offsets()` functions.